clippy_lints/src/slice_as_bytes.rs

```diff
@@ -0,0 +1,58 @@
+use clippy_utils::{diagnostics::span_lint_and_sugg, source::snippet_with_applicability, ty::is_type_lang_item};
+use rustc_errors::Applicability::MachineApplicable;
+use rustc_hir::{Expr, ExprKind, LangItem, QPath};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for string slices immediantly followed by `as_bytes`.
+    /// ### Why is this bad?
+    /// This involves doing an unnecessary UTF-8 alignment check which is less efficient, and can cause a panic.
+    /// ### Example
+    /// ```rust
+    /// let s = "Lorem ipsum";
+    /// s[1..5].as_bytes();
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let s = "Lorem ipsum";
+    /// &s.as_bytes()[1..5];
+    /// ```
+    #[clippy::version = "1.72.0"]
+    pub SLICE_AS_BYTES,
+    nursery,
+    "slicing a string and immediantly calling as_bytes is less efficient and can lead to panics"
+}
+declare_lint_pass!(SliceAsBytes => [SLICE_AS_BYTES]);
+
+impl<'tcx> LateLintPass<'tcx> for SliceAsBytes {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+        if let ExprKind::MethodCall(path, recv, [], _) = expr.kind && path.ident.name == sym!(as_bytes){
+            if let ExprKind::Index(indexed, index) = recv.kind {
+                use LangItem::{Range,RangeFrom,RangeFull,RangeTo,RangeToInclusive};
+                if let ExprKind::Struct(
+                    QPath::LangItem(Range | RangeFrom | RangeTo | RangeToInclusive | RangeFull, ..),
+                    ..,
+                ) = index.kind
+                {
+                    let ty = cx.typeck_results().expr_ty(indexed).peel_refs();
+                    let is_str = ty.is_str();
+                                        let is_string = is_type_lang_item(cx, ty, LangItem::String);
+                                        if  is_str|| is_string
+                    {
+                        let mut applicability = MachineApplicable;
+                        let stringish = snippet_with_applicability(cx,indexed.span,"..",&mut applicability);
+                        let range = snippet_with_applicability(cx,index.span,"..",&mut applicability);
+                        let sugg = format!("&{stringish}.as_bytes()[{range}]");
+                        let type_name = if is_str {"str"} else {"String"};
+                        let msg = format!("slicing a {type_name} before calling `as_bytes` results in needless UTF-8 alignment checks, and has the possiblity of panicking");
+                        let sp = expr.span;
+                        span_lint_and_sugg(cx,SLICE_AS_BYTES ,sp , &msg, "try",sugg ,applicability);
```

Comment:
No need to assign variables for superfluous values. Generally, we only store the suggestion when emitting a lint (if the suggestion is complex, otherwise don't). But this may be just personal taste

```suggestion
                        span_lint_and_sugg(cx,SLICE_AS_BYTES ,expr.span , format!("slicing a {type_name} before calling `as_bytes` results in needless UTF-8 alignment checks, and may panic"), "try",format!(&{stringish}.as_bytes()[{range}]), Applicability::MachineApplicable);
```

(If you apply this, format using `cargo dev fmt` after)